modules/k8s/service.go

```diff
@@ -0,0 +1,211 @@
+	endpoint, err := GetServiceEndpointFromClientE(clientset, service, servicePort)
+	require.NoError(t, err)
+	return endpoint
+}
+
+// GetServiceEndpointFromClientE will return the service access point using the following logic:
+// - For ClusterIP service type, return the URL that maps to ClusterIP and Service Port
+// - For NodePort service type, identify the public IP of the node (if it exists, otherwise return the bound hostname),
+//   and the assigned node port for the provided service port, and return the URL that maps to node ip and node port.
+// - For LoadBalancer service type, return the publicly accessible hostname of the load balancer.
+// - All other service types are not supported.
+func GetServiceEndpointFromClientE(clientset *kubernetes.Clientset, service *corev1.Service, servicePort int) (string, error) {
+	switch service.Spec.Type {
+	case corev1.ServiceTypeClusterIP:
+		// ClusterIP service type will map directly to service port
+		return fmt.Sprintf("%s:%d", service.Spec.ClusterIP, servicePort), nil
+	case corev1.ServiceTypeNodePort:
+		return findEndpointForNodePortService(clientset, service, int32(servicePort))
+	case corev1.ServiceTypeLoadBalancer:
+		ingress := service.Status.LoadBalancer.Ingress
+		if len(ingress) == 0 {
+			return "", NewKubernetesError("Load Balancer is not ready")
+		}
+		// Load Balancer service type will map directly to service port
+		return fmt.Sprintf("%s:%d", ingress[0].Hostname, servicePort), nil
+	default:
+		return "", NewKubernetesError("Unknown service type")
+	}
+}
+
+// Extracts a endpoint that can be reached outside the kubernetes cluster. NodePort type needs to find the right
+// allocated node port mapped to the service port, as well as find out the externally reachable ip (if available).
+func findEndpointForNodePortService(
+	clientset *kubernetes.Clientset,
+	service *corev1.Service,
+	servicePort int32,
+) (string, error) {
+	nodePort, err := findNodePort(service, int32(servicePort))
+	if err != nil {
+		return "", err
+	}
+	node, err := pickRandomNode(clientset)
+	if err != nil {
+		return "", err
+	}
+	nodeHostname, err := findNodeHostname(node)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s:%d", nodeHostname, nodePort), nil
+}
+
+// Given the desired servicePort, return the allocated nodeport
+func findNodePort(service *corev1.Service, servicePort int32) (int32, error) {
+	for _, port := range service.Spec.Ports {
+		if port.Port == servicePort {
+			return port.NodePort, nil
+		}
+	}
+	return -1, NewKubernetesError("Port %d is not a part of the service %s", servicePort, service.Name)
+}
+
+// pickRandomNode will pick a random node in the kubernetes cluster
+func pickRandomNode(clientset *kubernetes.Clientset) (corev1.Node, error) {
+	nodes, err := GetNodesByClientE(clientset, metav1.ListOptions{})
+	if err != nil {
+		return corev1.Node{}, err
+	}
+	if len(nodes) == 0 {
+		return corev1.Node{}, NewKubernetesError("There are no nodes in the cluster")
```

Comment:
What's the advantage of a single `NewKubernetesError` used throughout this code? It doesn't make it easy to type check for specific errors, unlike, for example, a `NoNodesInCluster` type that is specific to this exact error here.

Reply:
That's a good point. Will fix.